### ISSUE TITLE
Request matching with dynamic dates

### ISSUE THREAD

**Issue by andy-cross** [tomakehurst/wiremock#1475]
## Feature request

As discussed with Tom via the mailing list...

We're looking to be able to request match on a dynamic date.  We have a few calls which always have a date a number of days in the future from the present date.  So ideally we want to be able to use the date helper in the request matching.

e.g.
```
  "priority": 5,
  "request": {
    "method": "POST",
    "url": "/api/service/fetchPolicy",
    "bodyPatterns": [
      {
        "matchesXPath": "//*[local-name() = 'PolicyNumber' and @Val = 'pol123456']"
      },
      {
        "matchesXPath": "//*[local-name() = 'StartDate' and @Val = '{{now offset=\'30 days\' format=\'yyyy-MM-dd\'}}']"
      }
    ]
  }
```

Being able to use the full range of options within the date and time response templating helper to set the time/date in different formats, offset by days, hours, minutes etc would be ideal.

**Comment by harlalkap** [comment]
This will be very useful feature for our team also.

**Comment by HermanCodes** [comment]
Upvote, this feature would greatly help my team

**Comment by mtpsangabriel** [comment]
Need this as well it would be great to have this feature

**Comment by RecursionAversion** [comment]
This would be a useful feature for dynamic data!

**Comment by savagerob** [comment]
We could also make use of this feature

**Comment by tomakehurst** [comment]
I think this should support `before`, `after` and `between` operations, taking either absolute date(time)s or offsets from now. ISO8601 dates and any of the 3 valid RFC2616 formats should be automatically detected, otherwise you should be able to specify a format string. Format string can also be `epoch` or `unix` as with the Handlebars date helper.

Examples of how this should work:

```json
{
  "request": {
    "method": "GET",
    "urlPath": "/date-match",
    "headers": {
      "X-My-Date-1": {
        "before": "2021-05-01T00:11:22Z"
      },
      "X-My-Date-2": {
        "after": "04/07/2020",
        "format": "dd/MM/yyyy"
      },
      "X-My-Date-3": {
        "between": ["Tue, 15 Nov 1994 12:45:26 GMT", "Wed, 16 Nov 1994 10:00:01 GMT"]
      },
      "X-My-Date-4": {
        "before": {
          "nowOffset": "-3 days",
          "truncate": "day"
        }
      },
      "X-My-Date-5": {
        "between": [{
          "nowOffset": "-10 months",
          "truncate": "hour"
        }, {
          "nowOffset": "5 days",
          "truncate": "hour"
        }]
      }
    }
  }
}
```

With Java DSL equivalents obviously...

**Comment by tomakehurst** [comment]
Rules for zoned/local:

* If the expected value(s) are zoned and the incoming value is zoned then a match can be attempted.
* If the expected value(s) are zoned and the incoming value is local then a match cannot be attempted and will always return false.
* If the expected value(s) are local and the incoming value is local then a match can be attempted.
* If the expected value(s) are local and the incoming value is zoned then the expected values will be assumed to be UTC and matched on that basis.

**Comment by tomakehurst** [comment]
I've pushed a 1st cut of this to a branch. I've just done the `before` matcher so far but it does all the stuff described above (although I haven't sorted out JSON serialisation properly yet).

Examples here: https://github.com/tomakehurst/wiremock/blob/datetime-matcher/src/test/java/com/github/tomakehurst/wiremock/matching/DateTimePatternsTest.java

Feedback welcome.

**Comment by tomakehurst** [comment]
- [ ] Document matchers in request-matching.md
- [ ] Add to JSON schemas in OpenAPI spec
- [ ] Support custom date format with truncation in DSL
- [ ] Support epoch and unix formats

**Comment by tomakehurst** [comment]
OK, we have a final(ish) version of this on the main branch now.

Before I release it, I'd appreciate a review from those of you that requested/upvoted it to ensure it'll satisfy the use cases you had in mind.

You can find the docs here: https://github.com/tomakehurst/wiremock/blob/master/docs-v2/_docs/request-matching.md#dates-and-times

**Comment by andy-cross** [comment]
Thanks Tom, this looks great.

Just checking - the `between `now works as expected?

Is there a rought date planned for the release of 2.29.0?

**Comment by tomakehurst** [comment]
`between` is the one bit that's not done yet. You can get the same effect on request body attributes by adding multiple matchers of `before` and `after`.

However, you can't do the same thing for headers/cookies/query parameters at the moment.

The reason I'm holding off is that I'm experimenting with adding logical AND and OR matchers. The AND matcher would solve this problem, but also support a broader range of other use cases.

I'm hoping to try and get a release out sometime next week.

**Comment by harlalkap** [comment]
Hi Tom,

Is it possible to place wiremock standalone jar also in same branch so we
can give it a try for my project.

As we only use wiremock jar and we are behind corporate firewall, its
difficult for us to build amd get the jar.

Regards,
Pankaj

**Comment by tomakehurst** [comment]
@harlalkap here you go: https://github.com/tomakehurst/wiremock/releases/tag/2.29.0-beta

**Comment by tomakehurst** [comment]
- [ ] DSL should take Java 8 time objects, not just strings
- [ ] JSON format should take offset and unit as individual parameters, not just part of a string
- [ ] Logical OR diff message when used with absent isn't very clear

**Comment by harlalkap** [comment]
Will this date helper work with Xpath?

**Comment by tomakehurst** [comment]
Yes, you can use them as sub-matchers for xpath and jsonpath.

**Comment by harlalkap** [comment]
I tried to use with xpath but seems i am missing something.


  "priority": 5,
  "request": {
    "method": "POST",
    "url": "/api/service/fetchPolicy",
    "bodyPatterns": [
      {
        "matchesXPath": "//*[local-name() = 'PolicyNumber' and @Val =
'pol123456']"
      },
      {
        "matchesXPath": "//*[local-name() = 'StartDate' and @Val =
'{"before" : "now +3 days"}']"

      }
    ]

  }


On Mon, 28 Jun 2021 at 23:05, Tom Akehurst ***@***.***> wrote:

Yes, you can use them as sub-matchers for xpath and jsonpath.

**Comment by tomakehurst** [comment]
Rather than putting it in the actual XPath expression, you need to use the enhanced XPath matcher which allows you to attach a sub-matcher e.g.:

```json
{
  ...
  "bodyPatterns": [
    ...
    {
      "matchesXPath": {
        "expression": "//*[local-name() = 'StartDate']/@Val",
        "before": "now +3 days"
      }
    }
  ]
}
```

**Comment by andy-cross** [comment]
@tomakehurst I'm having issues getting this working with xpath, even after the advice above.

**Mapping**

```
	"priority": 5,
	"request": {
		"method": "POST",
		"url": "/services/testDates",
		"bodyPatterns": [{
			"matchesXPath": "//*[local-name() = 'Policy' and @Val = 'ABC123']"
		},
		{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
				"before": "now +3 days"
			}
		}]
	}
```

**Request**

```
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope>
    <soapenv:Header>
    </soapenv:Header>
    <soapenv:Body>
        <Retrieve>
            <Policy>
                <EffectiveDate Val="01/07/2021" />
                <Policy Val="ABC123" />
            </Policy>
        </Retrieve>
    </soapenv:Body>
</soapenv:Envelope>
```


If I remove the following then the mapping works as expected without the date matching....

```
{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
				"before": "now +3 days"
			}
		}
```

What am I doing wrong?  Haven't tried with jsonPath yet but will give it a look as soon as possible

**Comment by andy-cross** [comment]
@tomakehurst I'm having the same issue with jsonPath.

**Mapping**

```
{
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [{
      "matchesJsonPath": "$.searchCriteria[?(@.customerId == '104903')]",
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "equalToDateTime": "2021-01-01T00:00:00Z"
      }
    }]
```

**Request**

```
{
   "searchCriteria": {
      "customerId": "104903",
      "date": "2021-01-01T00:00:00"
   }
}
```

Again, if I remove the date matching it behaves as expected:

```
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "equalToDateTime": "2021-01-01T00:00:00Z"
      }
```

**Comment by tomakehurst** [comment]
In your XML example you're using a date format that's not one of the defaults, so you'd need to explicitly state the format you expect e.g. `"actualFormat" : "dd/MM/yyyy"`

In the JSON example you're mixing zoned and local dates - the expected is zoned, but the actual is local which means they  can't be compared. There's a note about this here: http://wiremock.org/docs/request-matching/#local-vs-zoned. The simplest way to fix this is to remove the `Z` from your expected date.

**Comment by andy-cross** [comment]
@tomakehurst thanks for this.  I got jsonpath working.  Not had a proper look at xpath yet - that's a tomorrow job so will let you know as soon as possible

**Comment by andy-cross** [comment]
@tomakehurst still struggling with xpath I'm afraid.  Are you able to offer any additional advice after adding in the `actualFormat`?

**Mapping**

```
{
	"priority": 5,
	"request": {
		"method": "POST",
		"url": "/services/testDates",
		"bodyPatterns": [{
			"matchesXPath": "//*[local-name() = 'Policy' and @Val = 'ABC123']"
		},
		{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
				"before": "now +3 days",
				"actualFormat" : "dd/MM/yyyy"
			}
		}]
	},
```

**Request**

```
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope>
    <soapenv:Header>
    </soapenv:Header>
    <soapenv:Body>
        <Retrieve>
            <Policy>
                <EffectiveDate Val="01/07/2021" />
                <Policy Val="ABC123" />
            </Policy>
        </Retrieve>
    </soapenv:Body>
</soapenv:Envelope>
```

**Comment by andy-cross** [comment]
@tomakehurst update:

My issues appear to be common across both xpath and jsonpath.  If I match on just `equalToDateTime` using the standard time format then it's working as expected

Mapping

```
{
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [{
      "matchesJsonPath": "$.searchCriteria[?(@.customerId == '104903')]",
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "equalToDateTime": "2021-01-01T00:00:00"
      }
    }]
  }
```

Request

```
{
   "searchCriteria": {
      "customerId": "104903",
      "date": "2021-01-01T00:00:00"
   }
}
```

However, if I add in an `actualFormat` and change the request as below, it's failing to find a match:

Mapping

```
{
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [{
      "matchesJsonPath": "$.searchCriteria[?(@.customerId == '104903')]",
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "equalToDateTime": "2021-01-01T00:00:00",
        "actualFormat" : "dd/MM/yyyy"

      }
    }]
  }
```

Request

```
{
   "searchCriteria": {
      "customerId": "104903",
      "date": "01/01/2021"
   }
}
```

Also, if I use a `before` or `after` it is failing to find a match:

{
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [{
      "matchesJsonPath": "$.searchCriteria[?(@.customerId == '104903')]",
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "before": "now + 3 days"
      }
    }]
  }
```

Request

```
{
   "searchCriteria": {
      "customerId": "104903",
      "date": "2021-01-01T00:00:00"
   }
}
```


This is common across both jsonpath and xpath using the standalone jar here - https://github.com/tomakehurst/wiremock/releases/tag/2.29.0-beta

**Comment by tomakehurst** [comment]
Thanks for sharing this. This has revealed a bug in `matchesJsonPath`: https://github.com/tomakehurst/wiremock/issues/1558.

I'll aim to fix this in 2.29.1.

**Comment by andy-cross** [comment]
@tomakehurst No worries.  Thanks for your help with all this.

As per my message above, we seem to have a common issue across `matchesJsonPath` and `matchesXPath` as we're seeing the same problems when using both `before` or `after` and with `actualFormat`.  Is it a common underlying cause?  https://github.com/tomakehurst/wiremock/issues/1558 only seems to cover `matchesJsonPath` for the `actualFormat` and not the `before` or `after` cases.  Just want to check that all issues are covered.

This matches fine:

**Mapping**

```
{
	"priority": 5,
	"request": {
		"method": "POST",
		"url": "/services/testDates",
		"bodyPatterns": [{
			"matchesXPath": "//*[local-name() = 'Policy' and @Val = 'ABC123']"
		},
		{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
        		        "equalToDateTime": "2021-01-01T00:00:00"
			}
		}]
	}
```

**Request**

```
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope>
    <soapenv:Header>
    </soapenv:Header>
    <soapenv:Body>
        <Retrieve>
            <Policy>
                <EffectiveDate Val="2021-01-01T00:00:00" />
                <Policy Val="ABC123" />
            </Policy>
        </Retrieve>
    </soapenv:Body>
</soapenv:Envelope>
```

However, this doesn't match:

**Mapping**

```
{
	"priority": 5,
	"request": {
		"method": "POST",
		"url": "/services/testDates",
		"bodyPatterns": [{
			"matchesXPath": "//*[local-name() = 'Policy' and @Val = 'ABC123']"
		},
		{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
        		        "before": "now +3 days"
			}
		}]
	}
```

**Request**

```
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope>
    <soapenv:Header>
    </soapenv:Header>
    <soapenv:Body>
        <Retrieve>
            <Policy>
                <EffectiveDate Val="2021-01-01T00:00:00" />
                <Policy Val="ABC123" />
            </Policy>
        </Retrieve>
    </soapenv:Body>
</soapenv:Envelope>
```

**Mapping**

```
{
	"priority": 5,
	"request": {
		"method": "POST",
		"url": "/services/testDates",
		"bodyPatterns": [{
			"matchesXPath": "//*[local-name() = 'Policy' and @Val = 'ABC123']"
		},
		{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
				"equalToDateTime": "2021-01-01T00:00:00",
				"actualFormat" : "dd/MM/yyyy"
			}
		}]
	}
```

**Request**

```
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope>
    <soapenv:Header>
    </soapenv:Header>
    <soapenv:Body>
        <Retrieve>
            <Policy>
                <EffectiveDate Val="01/1/2021" />
                <Policy Val="ABC123" />
            </Policy>
        </Retrieve>
    </soapenv:Body>
</soapenv:Envelope>
```

**Comment by tomakehurst** [comment]
The issue is to do with how `matchesXPath` and `matchesJsonPath` apply sub-matchers (the problem is mostly in a common base class), so it will affect any of these, not just the date-oriented ones.

I think I've pretty much fixed it now, barring a few more tests I want to write around `matchesXPath` specifically.

**Comment by tomakehurst** [comment]
@andy-cross I've pushed a fix for this now and I've added test cases using the examples you've provided above. One minor tweak was using double-digit months in the XML date value i.e. `01/01/2021` rather than `01/1/2021`: https://github.com/tomakehurst/wiremock/blob/master/src/test/java/com/github/tomakehurst/wiremock/matching/MatchesXPathPatternTest.java#L186

**Comment by andy-cross** [comment]
Thanks for this @tomakehurst.  Is it possible to get a standalone jar with the fix in please?  Much like one of the earlier commenters, my corporate firewall means there's a bunch of hoops to jump through to get and build myself.

**Comment by tomakehurst** [comment]
https://github.com/tomakehurst/wiremock/releases/tag/2.29.1-beta

**Comment by tomakehurst** [comment]
@andy-cross FYI I'm going to try and ship this before the weekend so if you've got any feedback I'll be able to incorporate it more quickly before then.

**Comment by andy-cross** [comment]
@tomakehurst thanks for the info - it's on my to-do list for this afternoon.  I'll be in touch

**Comment by andy-cross** [comment]
@tomakehurst `actualFormat` pattern match now working as expected across both `matchesXPath` and `matchesJsonPath`.

However, I've still not got the `before` and `after` matching working for either  `matchesXPath` or `matchesJsonPath`.

I've tried this both with and without `actualFormat`.

None of the below work...

### xPath

**Mapping**

```
{
	"priority": 5,
	"request": {
		"method": "POST",
		"url": "/services/testDates",
		"bodyPatterns": [{
			"matchesXPath": "//*[local-name() = 'Policy' and @Val = 'ABC123']"
		},
		{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
				"before": "now +3 days"
			}
		}]
	}
```

**Request**

```
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope>
    <soapenv:Header>
    </soapenv:Header>
    <soapenv:Body>
        <Retrieve>
            <Policy>
                <EffectiveDate Val="2021-07-01T00:00:00" />
                <Policy Val="ABC123" />
            </Policy>
        </Retrieve>
    </soapenv:Body>
</soapenv:Envelope>
```

**Mapping**

```
{
	"priority": 5,
	"request": {
		"method": "POST",
		"url": "/services/testDates",
		"bodyPatterns": [{
			"matchesXPath": "//*[local-name() = 'Policy' and @Val = 'ABC123']"
		},
		{
			"matchesXPath": {
				"expression": "//*[local-name() = 'EffectiveDate']/@Val",
				"before": "now +3 days",
				"actualFormat" : "dd/MM/yyyy"
			}
		}]
	}
```

**Request**

```
<?xml version="1.0" encoding="utf-8"?>
<soapenv:Envelope>
    <soapenv:Header>
    </soapenv:Header>
    <soapenv:Body>
        <Retrieve>
            <Policy>
                <EffectiveDate Val="01/07/2021" />
                <Policy Val="ABC123" />
            </Policy>
        </Retrieve>
    </soapenv:Body>
</soapenv:Envelope>
```

### jsonPath

**Mapping**

```
{
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [{
      "matchesJsonPath": "$.searchCriteria[?(@.customerId == '104903')]",
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "before": "now +3 days"
      }
    }]
  }
```

**Request**

```
{
   "searchCriteria": {
      "customerId": "104903",
      "date": "2021-07-01T00:00:00"
   }
}
```

**Mapping**

```
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [{
      "matchesJsonPath": "$.searchCriteria[?(@.customerId == '104903')]",
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "before": "now +3 days",
        "actualFormat" : "dd/MM/yyyy"
      }
    }]
  }
```

**Request**

```
{
   "searchCriteria": {
      "customerId": "104903",
      "date": "01/07/2021"
   }
}
```

**Comment by tomakehurst** [comment]
This is failing because an offset expression will produce a zoned date/time whereas you're comparing with a local one.

I'd concluded that it was better to not match zoned -> local in this way because it's potentially misleading i.e. would produce different results depending on the timezone of the local machine.

But I can see this is going to come up a lot, so maybe doing the comparison by assuming a timezone for the actual date is the right way to go. The question is should that assumed timezone by UTC or the machine's local timezone?

**Comment by tomakehurst** [comment]
I've plumped for using the system timezone, since this is consistent with how offset dates are yielded.

**Comment by tomakehurst** [comment]
New build here: https://github.com/tomakehurst/wiremock/releases/tag/2.29.1-beta2

**Comment by andy-cross** [comment]
@tomakehurst thanks for this, will take a look now.

**Comment by andy-cross** [comment]
@tomakehurst taken a look at this and it seems to be behaving exactly as expected for `matchesXPath`

However, with `matchesJsonPath` I'm seeing odd behaviour.  I'm matching on 2 criteria `customerId` and `date`.

**Mapping**

```
{
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [{
      "matchesJsonPath": "$.searchCriteria[?(@.customerId == 'before')]",
      "matchesJsonPath": {
        "expression": "$.searchCriteria.date",
        "before": "now +30 days",
        "actualFormat" : "dd/MM/yyyy"
      }
    }]
  }
```

With the following request it matches as expected:

```
{
   "searchCriteria": {
      "customerId": "before",
      "date": "01/07/2021"
   }
}
```

However, the following is also matching so it seems like the `customerId` matching criteria is being ignore and it's just looking at the dates

```
{
   "searchCriteria": {
      "customerId": "andytest",
      "date": "01/07/2021"
   }
}
```

**Comment by tomakehurst** [comment]
I think it's because you've put two `matchesJsonPath` elements in the same object, whereas they need to be separate objects in the array. Try this:

```json
{
  "priority": 10,
  "request": {
    "urlPattern": "/api/dates",
    "method": "POST",
    "bodyPatterns": [
      {
        "matchesJsonPath": "$.searchCriteria[?(@.customerId == 'before')]"
      },
      {
        "matchesJsonPath": {
          "expression": "$.searchCriteria.date",
          "before": "now +30 days",
          "actualFormat": "dd/MM/yyyy"
        }
      }
    ]
  }
}
```

**Comment by andy-cross** [comment]
@tomakehurst of course, thanks!  Both `matchesJsonPath` and `matchesXPath` are working as expected now.

Thanks for your help with this one

**Comment by tomakehurst** [comment]
I've released this under 2.29.1 now, so I'll close it off.

Thanks @andy-cross for all the feedback!
